frontend/plan/src/syncutils.js

```diff
@@ -0,0 +1,101 @@
+import {
+    attemptDeletion,
+    createScheduleOnBackend,
+    deleteSchedule,
+    deleteScheduleOnBackend,
+    deletionAttemptCompleted,
+    deletionSuccessful,
+    fetchSchedulesAndInitializeCart,
+    incrementDeletionAttempts,
+    updateScheduleOnBackend
+} from "./actions";
+import fetch from "cross-fetch";
+import getCsrf from "./csrf";
+
+// Ensure that fetches only happen once ever 250 ms
+let lastFetched = 0;
+export const conditionalFetch = (url, init) => {
+    const now = Date.now();
+    if (now - lastFetched > 250) {
+        return fetch(url, init);
+    }
+    return null;
+};
+
+/**
+ * Initiates schedule syncing on page load
+ * @param store The redux store
+ */
+const initiateSync = store => {
+
+    // Retrieve all the schedules that have been observed coming from
+    // the backend at any point in time.
+    // This ensures that schedules can be safely deleted without coming back.
+    let schedulesObserved = {};
+    const localStorageSchedulesObserved = localStorage.getItem("coursePlanObservedSchedules");
+    if (localStorageSchedulesObserved) {
+        try {
+            schedulesObserved = JSON.parse(localStorageSchedulesObserved);
+        } catch (ignored) {
+        }
+    }
+
+    const scheduleStateInit = store.getState().schedule;
+    store.dispatch(fetchSchedulesAndInitializeCart(scheduleStateInit.cartSections, newSchedulesObserved => {
+        // record the new schedules that have been observed
+        newSchedulesObserved.forEach(({ id }) => {
+            schedulesObserved[id] = true;
+        });
+        Object.keys(schedulesObserved)
+            .forEach(id => {
+                // The schedule has been observed from the backend before, but is no longer being observed.
+                //Should be deleted locally
+                if (!newSchedulesObserved[id]) {
+                    delete schedulesObserved[id];
+                    // find the name of the schedule with the deleted id
+                    let schedName = Object.keys(scheduleStateInit.schedules)
+                        .reduce((acc, schedNameSelected) => acc || ((scheduleStateInit
+                            .schedules[schedNameSelected].id === id) && schedNameSelected), false);
+                    if (schedName) {
+                        store.dispatch(deleteSchedule(schedName));
+                    }
+                }
+            });
+        localStorage.setItem("coursePlanObservedSchedules", JSON.stringify(schedulesObserved));
+        window.setInterval(() => {
```

Comment:
Should make it a bit more clear that this is setting up a timer that keeps running repeatedly. Took me a bit to see that